Add cv-qualifiers to Callback and deprecate combinatorial explody functions

geky opened the pull request:
From this discussion https://github.com/ARMmbed/mbed-os/pull/2395 and the related issue https://github.com/ARMmbed/mbed-os/issues/2490

The combination of cv-qualifiers with callback overloads results in a large number of overloads to match all possible types that could be passed to the Callback class.

This isn't a big problem for `void (*)(cv T *)` function types, but for `void (T::*)() cv` types, the method in which cv-qualifiers are associated prevents encoding the cv-qualifiers in the `T` type, and requiring a large combination of overloads for all cv-qualifiers.

This pr involves several changes:
- Adds full support for cv-qualifiers in Callback constructor
- Provides convenience `mbed::callback` function to pass different arguments to Callback with infered type.
- Deprecates existing convenience functions to avoid confusion if the user attempts to use a `const` or `volatile` object with the existing convenience functions.
- Updates tests appropriately

Supported overloads:
``` cpp
callback(void (*f)(A...));
callback(const Callback<R(A...)> &);
callback(T *t,                void (*f)(T*, A...));
callback(const T *t,          void (*f)(const T*, A...));
callback(volatile T *t,       void (*f)(volatile T*, A...));
callback(const volatile T *t, void (*f)(const volatile T*, A...));
callback(T *t,                void (T::*f)(A...));
callback(const T *t,          void (T::*f)(A...) const);
callback(volatile T *t,       void (T::*f)(A...) volatile);
callback(const volatile T *t, void (T::*f)(A...) const volatile);
```

Should fix: https://github.com/ARMmbed/mbed-os/issues/2490
Replaces:  https://github.com/ARMmbed/mbed-os/pull/2395 and https://github.com/ARMmbed/mbed-os/pull/2495
cc @pan-, @0xc0170

Let me know any thoughts/suggestions on this solution you guys have.

pan- commented:
That's a really nice change, I have few comments to make: 

* It would be interesting to make `call` and `operator()` `const` like in `std::function`, it allow user code to pass callback around with the guarantee that it will not be modified. 
* The `operator=` is missing. 
* Maybe it would be interesting to add external operators `==` and `!=`.
* The `call` function might not be instantiated if `R` can't be constructed or casted from the literal `0` (see [here](https://github.com/geky/mbed/blob/29bda2e58edbbfcd58a62aaebaebd6acaf543fef/hal/api/Callback.h#L246-L248)). 
* Please provide relevant comments for every overload and template class specialization. It will help a lot with online documentation. Otherwise, user might get nothing if the overload is not documented or not useful comment like `Templated function class`.

geky commented:
+1, these are good points

Since a copy involves a simple copy of all the members, is the default `operator=` not sufficient?

Unfortunately the cast is for backwards compatibility, since at some point someone thought `FunctionPointer f; f();` should be well defined. I can update it to value initialize `R` without breaking backwards compatibility, although that still assumes more about `R` than should be necessary.

pan- commented:
> Since a copy involves a simple copy of all the members, is the default operator= not sufficient?

Yes it is sufficient but in that case drop the copy construction operator (rule of 0 or rule of 3 (five in c++11)). 

> Unfortunately the cast is for backwards compatibility, since at some point someone thought FunctionPointer f; f();

The question is what we should do in that case ? 
Instead of returning something totally invalid maybe it can be trapped into an error. 

```c++
    R call() {
        if (!_thunk) {
            // never returns
            error("LOGIC_ERROR: Call to a not initialized callback!");
        }
        return _thunk(_obj, &_func);
    }
```

So `FunctionPointer f; f();` is still valid, the user is notified in such case and there is template instantiation issues.

geky commented:
The reason for explicitely defining the copy constructor was to make it explicit that the Callback class could be copied safely. Tangentally do we expect this to work? Might need the assignment operator:
``` cpp
Callback<void()> cb;
void doit();
cb = doit;
```

(Still looking into where null callback is used)

geky commented:
Well I can't seem to find where the null callback was relied on, I thought it was somewhere in our codebase. I went ahead and moved the null check into th `e deprecated `FunctionPointer` class to maintain that backwards compatibility. I tested a few things locally and we'll see if tests find any issues.

Interestingly enough, the `operator=` seems to work fine with C style function pointers at the moment, I think this is because it infers the cast before the assignment. I went ahead and added a test for that.

I'd prefer to keep the copy constructor since it clarifies the class can be safely copied and matches the overloads for `attach` and `callback`, though I can remove it if we really want it gone.

geky commented:
Looks like it is a bit tricky, since several drivers are relying on uninitialized Callbacks being callable, however this is an easy fix.

pan- commented:
> Interestingly enough, the operator= seems to work fine with C style function pointers at the moment, I think this is because it infers the cast before the assignment. I went ahead and added a test for that.

`Callback` function pointer constructors are not tagged with explicit and therefore act as conversion constructors (more [here](http://en.cppreference.com/w/cpp/language/converting_constructor)). This is the expected behavior. 

> I'd prefer to keep the copy constructor since it clarifies the class can be safely copied and matches the overloads for attach and callback, though I can remove it if we really want it gone.

@0xc0170 @sg- Do we recommend any pattern like `rule of three` or `rule of zero` for C++ value types ? 
What is the best approach here in term of readability and *developer experience* ? 

Otherwise, +1 for the whole PR.

0xc0170 commented:
> @0xc0170 @sg- Do we recommend any pattern like rule of three or rule of zero for C++ value types ?
What is the best approach here in term of readability and developer experience ? 

+1, this should be covered in the guidelines. I like a rule "all or nothing". We already provide "rule of three" or at least partially in our classes, we manage resources, thus inclined to rule "all", for us (c++03) three. If we agree, ``Callback`` should follow it.

The changes look good. People happy with jinja2 generated files?

sg- commented:
> I like a rule "all or nothing"

👍 

> The changes look good. People happy with jinja2 generated files?
nope

geky commented:
@sg-, is there an alternative template engine we should be using? [m4](https://en.wikipedia.org/wiki/M4_(computer_language)) perhaps?

Or do you suggest just not committing the template files?

sg- commented:
Not committing the template files. If something needs to change that often to need auto-generation, I don't think its ready for master :)

geky commented:
Copy contructor and jinja templates removed, let me know any other feedback.

0xc0170 commented on TESTS/mbed_drivers/callback/main.cpp:
```diff
@@ -1,3 +1,6 @@
+// Note: This file was autogenerated from main.j2
+// j2 main.j2 > main.h
```
This is not true anymore?

geky replied:
Good catch!

0xc0170 commented:
LGTM

geky commented:
Should be good to go :+1:

geky commented:
/morph test

geky commented:
Looks like all the template expansions in the test caused flash to exceed 64K : /

After talking with @bridadan (correct me if I'm wrong), will split the test into parts and move to TESTS/mbed_functional, which can house other non-hardware apis.

geky commented:
/morph test

pan- commented:
> Looks like all the template expansions in the test caused flash to exceed 64K : /

At a latter stage maybe we can put all the [data members](https://github.com/geky/mbed/blob/34f0bb82d643a314f0e1adb6500e46f982de1a65/hal/api/Callback.h#L346-L356) in a non template base class and have function to set and reset them.

That way `operator bool` can be non template and [reset and assignment](https://github.com/geky/mbed/blob/34f0bb82d643a314f0e1adb6500e46f982de1a65/hal/api/Callback.h#L297-L300) can be non template.

It is also possible to factorize all `operator==` and `operator!=` instantiation, it is just a memcmp of this base class.

sg- commented:
@geky please squash this into 2 or 3 commits:
- Added support for cv-qualifiers in Callback class
- Removed copy-constructor from callback constructor
- Split callback test into multiple tests based on types or a addition of tests commit

geky commented:
/morph test

bridadan commented:
@geky Sorry, making some CI changes. Please hold off on firing more PRs for the moment

geky commented:
/morph test

bridadan commented:
Looks like there was bug introduced with the latest version of setuptools, oh joy! I've deployed a workaround that should fix the test prep issue for now until they patch setuptools.

/morph test

geky commented:
/morph test

bridadan commented:
@geky Test failures look ok :+1: Just timing issues again from the CI machine being under load

geky commented:
Ah ok, in that case this pr should be good to go

LGTM?

pan- commented:
LGTM ?

teetak01 commented:
Hi, this broke all mbed Client components. Please revert this.

https://jenkins-internal.mbed.com/view/mbed_Client/job/mbed-os-example-client_master/label=gcc,target=morpheus/114/console

yogpan01 commented:
@sg- @geky @bridadan Why did you ignore https://github.com/ARMmbed/mbed-os/issues/2653 because this PR has now broken the whole mbed-client builds ?
Please revert this PR ASAP or push the fix to resolve this. All our CI build jobs for mbed CLient are now broken and team is blocked because of this.
@jupe @marhil01 @tommikas FYI

jupe commented:
we need also this https://github.com/ARMmbed/mbed-os/pull/2656 to be merged asap because it could find this kind of issues.. but I'm wondering why any other CI job's didn't recognize this before it was too late..?

yogpan01 commented:
Our nightly build job identified this issue , it has been failing since yesterday morning after this PR has been merged.

jupe commented:
i think this kind of issues should be identified during PR verification jobs - before merging..

yogpan01 commented:
mbed-os needs to add mbed-client build jobs to the PR verification queue.

sg- commented:
> @sg- @geky @bridadan Why did you ignore #2653 because this PR has now broken the whole mbed-client builds ?
Please revert this PR ASAP or push the fix to resolve this. All our CI build jobs for mbed CLient are now broken and team is blocked because of this.
@jupe @marhil01 @tommikas FYI

Will patch or revert. Wasnt aware of this issues as it wasnt linked to the pr until after the merge

sg- commented:
> i think this kind of issues should be identified during PR verification jobs - before merging..

There is no verification job on that catches this. They're missing from the jenkinsfile

sg- commented:
> we need also this #2656 to be merged asap because it could find this kind of issues.. but I'm wondering why any other CI job's didn't recognize this before it was too late..?

This has failing status so I cant merge :-1:
